File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int maxArea = 0;

        for (int i = 0; i < heights.size(); i++) {
            int cur = heights[i];
            int left = i - 1;
            int right = i + 1;
            int count = 1;

            // Expand left
            while (left >= 0 && heights[left] >= cur) {
                count++;
                left--;
            }

            // Expand right
            while (right < heights.size() && heights[right] >= cur) {
                count++;
                right++;
            }

            maxArea = std::max(maxArea, cur * count);
        }

        return maxArea;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l, r = 0, len(heights)-1
        area = 0
        while l <= r:
            a = heights[l]
            b = heights[r]

            area = max(area, (r-l) * min(a, b))
            if min(a, b) == a:
                l += 1
            else:
                r -= 1
        return area